### ISSUE TITLE
Lazy Loading for parentTypeFns

### ISSUE THREAD

**Issue by DellanX** [dipscope/TypeManager.TS#33]
**Is your feature request related to a problem? Please describe.**

I'm running into some headache with my code, as I have a circular definition, which seems to really make this library upset.
Specifically, I have an "interface" class `messageable` which indicates that that the implementing class will have the `messages` property. (An EntityCollection of type `message`). Unfortunately, messages are also `messageable`, as they can have sub-threads attached. Unless I do my imports in a very specific way, I run into the following error:

```log
TypeError: Cannot read properties of undefined (reading 'prototype')
```

**Describe the solution you'd like**

I have no idea how the decorators are handling parentTypeFns; however, I was curious if it'd be possible to have a lazy-loader path for these definitions? If not, I imagine I'll need to use the other syntax to define this type's metadata.

**Describe alternatives you've considered**

I used to due some bulk imports, which worked somehow. I may go back to that for these types. While I am moving away to optimize my bundle, I think I could make an exception here.

**Additional context**

The full error log

```log
TypeError: Cannot read properties of undefined (reading 'prototype')
 ❯ extractTypeMetadata node_modules/@dipscope/type-manager/dist/umd/webpack:/$DSTypeManager/src/type-manager.ts:490:34
 ❯ n node_modules/@dipscope/type-manager/dist/umd/webpack:/$DSTypeManager/src/type-metadata.ts:1123:57
 ❯ t.hasParentTypeFns node_modules/@dipscope/type-manager/dist/umd/index.js:1:106119

 ❯ t.configure node_modules/@dipscope/type-manager/dist/umd/webpack:/$DSTypeManager/src/type-metadata.ts:1153:18
 ❯ e.configureTypeMetadata node_modules/@dipscope/type-manager/dist/umd/webpack:/$DSTypeManager/src/type-manager.ts:434:26
 ❯ Function.configureTypeMetadata node_modules/@dipscope/type-manager/dist/umd/webpack:/$DSTypeManager/src/type-manager.ts:404:39
 ❯ node_modules/@dipscope/type-manager/dist/umd/webpack:/$DSTypeManager/src/type.ts:20:42
 ❯ __decorateClass src/composables/models/chats/message.ts:1435:144
 ❯ src/composables/models/chats/message.ts:5:1
```

**Comment by dpimonov** [comment]
Hi [DellanX](https://github.com/DellanX),

Thank you, as always, for your detailed explanation. The issue with inheritance when using decorators arises because decorators are constructed once and then reused. This is particularly problematic with parent classes, as their properties must be declared before those of child classes. This ensures that potential overrides, which are not currently handled by the library, work as expected.

If overrides are not required in your case, you can declare the `parentTypeFns` later using a declarative configuration approach. This works because the class and its entity configurations are declared separately, breaking the circular dependency. You can apply this declarative style specifically to the model causing the issue.

Here’s an example:

```typescript
@Type()
class Messageable 
{
    @Property(EntityCollection, [() => Message]) public messages: EntityCollection<Message>;
}

@Type({
    parentTypeFns: [Messageable]
})
class Message implements Messageable 
{
    // ...
}
```

would be refactored as:

```typescript
// This part remains unchanged.
@Type()
class Messageable 
{
    @Property(EntityCollection, [() => Message]) public messages: EntityCollection<Message>;
}

// Avoid declaring parentTypeFns here.
@Type()
class Message implements Messageable 
{
    // ...
}

// Create a separate configuration class to resolve the circular dependency.
class MessageConfiguration implements TypeConfiguration<Message> 
{
    public configure(typeMetadata: TypeMetadata<Message>): void 
    {
        typeMetadata.hasParentTypeFns([Messageable]);

        return;
    }
}
```

Later, during app initialization or at any point where code runs only once, apply the configuration:

```typescript
TypeManager.applyTypeConfiguration(Message, new MessageConfiguration());
```

In general, I recommend using declarative configurations through `TypeConfiguration` classes, especially from TypeScript v5 onwards. With the recent refactor of decorators, certain features like constructor injection are no longer possible under the new decorator syntax as for now.

Regarding the circular inheritance issue, this is definitely something that needs to be addressed in the future. However, it requires significant changes to the library to properly support lazy initialization of parent classes while maintaining the functionality of child property overrides. This limitation is the primary reason lazy initialization is not currently feasible. We can keep this issue open as a reminder. Let me know if the above solution works for you.

**Comment by DellanX** [comment]
Oh man, I didn't realize I could do that!... I may change my approach then, and structure my library similar to how a C++ would be set up.

That said, I am about 80% done with model unit testing. Once I am done with setting up my unit tests, I may migrate to `TypeConfiguration` classes. I don't use constructor injection but want to migrate for stability purposes. (Also I want to see if it'll work on `NUXT`, which does not like my current config)

As always, thank you for the information! As for my side of things, I have a few paths forward and am comfortable if you want to close this. My current solution is to just have both model files in the same file.

As always, thank you for the information, you have been very helpful!
And always worth restating, this library is great!

**Comment by dpimonov** [comment]
I believe that TypeScript, in general, does not handle circular references well.

```
My current solution is to just have both model files in the same file.
```

It is also possible to use aliases to resolve circular dependencies, especially when working with decorators. This approach avoids direct references to the class.

```typescript
import { Type, Property } from '@dipscope/type-manager';

@Type({
    alias: 'UserStatus',
})
export class UserStatus 
{
    @Property(String) public title: string;
}

@Type()
export class User 
{
    @Property('UserStatus') public userStatus: UserStatus;
}
```

```
I have a few paths forward and am comfortable if you want to close this.
```

Currently, I am working on a solution that allows for lazy inheritance, which seems to address the issue you’re facing. My focus has been on performance optimization, which has always been challenging due to the many configuration options and extensions involved. Fortunately, I’ve found a solution that not only optimizes performance but also resolves this circular dependency issue. So I will post there a message when fix will be available.

**Comment by DellanX** [comment]
> I believe that TypeScript, in general, does not handle circular references well.

no, no it does not. I started migrating away from declarators and found that most my tests still fail due to TypeScript issues.

Thanks for the info, I might start transitioning to using aliases before removing declarators, as I have discovered that migrating to type configurations is not a simple process. (Especially when I have a bunch of models).

> with the recent refactor of decorators, certain features like constructor injection are no longer possible under the new decorator syntax as for now.

Luckily, I don't ever use injectors, so I won't be affected by that portion. But perhaps I will have migrated away from declarators by the time I need to migrate to the new TypeScript Declarators.

**Comment by dpimonov** [comment]
Today, I completed the changes I mentioned last time. You can continue using the decorator approach since the entire configuration is now lazy. The `parentTypeFns` option has been renamed to `parentTypeArguments` and functions similarly to the one you provide in the `Property` config. It can now directly accept types, lazy function resolvers, or aliases.

There were significant changes in the latest release. While all tests have passed, I’m not entirely confident about its stability for updating `EntityStore` and `JsonApiEntityProvider` just yet. I plan to conduct more tests on a real API when I have time. I’ll post an update once it’s ready. 

In the meantime, if you’re interested, you can try using the latest version locally or assist with writing more specific tests for `TypeManager`. Your entity configurations are complex and well-designed, which I really like. :)

**Comment by DellanX** [comment]
Looks like JsonApiEntityProvider keys off of some removed functionality:

```log
tests/components/core/elements/autocomplete.test.ts
TypeError: this.typeMetadata.hasCustomOption is not a function
 - /node_modules/@dipscope/json-api-entity-provider/dist/umd/webpack:/$DSJsonApiEntityProvider/src/json-api-resource-metadata.ts:58:27
 - /node_modules/@dipscope/json-api-entity-provider/dist/umd/webpack:/$DSJsonApiEntityProvider/src/json-api-resource-metadata.ts:91:18
 - /node_modules/@dipscope/type-manager/dist/umd/webpack:/$DSTypeManager/src/type-extension-metadata.ts:29:14
 - /node_modules/@dipscope/json-api-entity-provider/dist/umd/webpack:/$DSJsonApiEntityProvider/src/json-api-resource-metadata.ts:24:9
 - /node_modules/@dipscope/type-manager/dist/umd/webpack:/$DSTypeManager/src/type-metadata.ts:519:39
 - /node_modules/@dipscope/json-api-entity-provider/dist/umd/webpack:/$DSJsonApiEntityProvider/src/json-api-resource.ts:19:51
 - /src/composables/data/types/identifiable.ts:511:141
 - /src/composables/data/types/identifiable.ts:541:16
 ```

Here in the next few days, I can see about testing the library a bit more

**Comment by dpimonov** [comment]
Yeap, it was replaced with `hasCustomValue` method in v8.

**Comment by DellanX** [comment]
I've confirmed this works, and it works really well/fast. Thanks!
